sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,681 @@
+following rules must be observed:
+
+* The position of the macro decoration must be the same as the position allowed
+  for C++ attribute decorations on the function declaration.
+  For example:
++
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr);
+```
+
+* The macro decoration must appear on the first declaration of the function
+  in the translation unit.
+  Redeclarations of the function may optionally be decorated with the same
+  macro if the macro arguments are the same.
+  The effect is the same regardless of whether redeclaration are so decorated.
+
+* The same function may be decorated with at most one of these macros.
+  However, the same function may be decorated multiple times with the same
+  macro (with the same arguments).
+  Programs that decorate the same function with more than one of these macros
+  or with multiple instances of the same macro with different arguments are ill
+  formed.
+
+* If a function is decorated with one of these macros in one translation unit,
+  that function must also be decorated with the same macro (with the same
+  arguments) in all other translation units.
+
+The following table provides additional details about these macros:
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+#define SYCL_EXT_ONEAPI_RANGE_KERNEL(Dims) /*unspecified*/
+----
+!====
+
+Indicates that the function is a free function kernel that is invoked with a
+simple `range` iteration space of `Dims` dimensions.
+The `Dims` parameter must be an integral constant expression that specifies a
+valid number of dimensions for the `range` class.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+#define SYCL_EXT_ONEAPI_ND_RANGE_KERNEL(Dims) /*unspecified*/
+----
+!====
+
+Indicates that the function is a free function kernel that is invoked with an
+`nd_range` iteration space of `Dims` dimensions.
+The `Dims` parameter must be an integral constant expression that specifies a
+valid number of dimensions for the `nd_range` class.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+#define SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL /*unspecified */
+----
+!====
+
+Indicates that the function is a free function kernel that is invoked via
+`single_task` (i.e. without any iteration space).
+|====
+
+When a function is defined as a free function kernel, each parameter to the
+function is a kernel argument, which must abide by the rules for allowable
+kernel parameter types specified in section 4.12.4 "Rules for parameter passing
+to kernels" of the core SYCL specification.
+The function's return type must be `void`.
+
+The following example demonstrates how a free function kernel using a
+3-dimensional nd-range iteration space can be defined:
+
+```
+SYCL_EXT_ONEAPI_ND_RANGE_KERNEL(3)
+void iota(float start, float *ptr) {
+   // ...
+}
+```
+
+A function decorated with one of these macros can still be called from host
+code.
+The macro has no effect in such cases.
```

Comment:
Yes, that's a good point.  I'll update the spec to allow this.

BTW, this relates to a question @AlexeySachkov raised.  In SPIR-V, it's illegal to call to a kernel function from anywhere else in the SPIR-V module.  We can still allow kernel functions to be called at the SYCL level, though.  It just means that the compiler needs to generate a wrapper around the kernel function.  That way the wrapper become the kernel function  at the SPIR-V level, and there will be no calls to this wrapper from the application.

I believe our device compiler already generates a wrapper like this, so I think this will fall out for free.

Reply:
I fixed this in 73ddbad24e765ad1db7d136d532818b8df55a80a.